### ISSUE TITLE
How can I control transparency in vispy?

### ISSUE THREAD

**Issue by skostrichegg** [vispy/vispy#1343]
Hi,

I've searched how to control transparency in vispy.
But I can't find it...

If I draw two figures by using scene.visuals.Mesh().
And I adjust camera angle to make them overlap.
(The back one is covered by the front one.)

I want to control transparency of the front one to showing back one.

How can I control transparency?,,

Thank you.

**Comment by kmuehlbauer** [comment]
@skostrichegg 

You can attach an alpha-filter to the front object. Have a look at this example how filters are used:

https://github.com/vispy/vispy/blob/master/examples/basics/visuals/visual_filters.py

**Comment by skostrichegg** [comment]
@kmuehlbauer Thank you for response.
Which scipy version has visuals.filters? I'm using versoin 0.4.0, but when I tried with "from vispy.visuals.filters import Alpha", it's failed to import(No module 'filters').
So, I tried with "from vispy.visuals.components.color2 import Alpha" instead of it.

from vispy.visuals.components.color2 import Alpha
mesh = scene.visuals.Mesh(...)
mesh.attach(Alpha(0.2))

However, looks like mesh's color tone is down, and it is not transparent. (not showing back one)

**Comment by kmuehlbauer** [comment]
@skostrichegg You need to use latest github version of vispy. Version 0.4.0 is heavily outdated. We pushing hard to get all those changes into 0.5.0 version soon (#1322). Sorry for any discomfort.

Can you show a minimum working example of your code? It would be much easier then to go into details.

**Comment by skostrichegg** [comment]
@kmuehlbauer OK, I installed vispy latest version. But front one still is not transparent.. I attach my example codes.

import numpy as np
from vispy import io, scene
from vispy.visuals import transforms
from vispy.visuals.filters import Alpha

# Canvas, view, camera
canvas = scene.SceneCanvas(bgcolor='w', size=(1250, 750), fullscreen=True, dpi=100)
view = canvas.central_widget.add_view()
camera = scene.TurntableCamera(elevation=80, azimuth=0)
camera.set_range((-44, 0), (-65, 43), (-26, 48))
view.camera = camera


# 1. Front one (Mesh)
n_radii = 8
n_angles = 36

radii = np.linspace(-40, 0, n_radii)
angles = np.linspace(0, 2*np.pi, n_angles, endpoint=False)

angles = np.repeat(angles[..., np.newaxis], n_radii, axis=1)

x = np.append(0, (radii*np.cos(angles)).flatten())
y = np.append(0, (radii*np.sin(angles)).flatten())

z = np.sin(-x*y)

mesh = scene.visuals.Mesh(np.vstack((x, y, z)).T, color=(0, 1, 0),\
        parent=view.scene)

mesh.attach(Alpha(0.2))


# 2. Back one (2D Ellipse)
pos = (-10, -10, 0)
el = scene.visuals.Ellipse(center=pos,\
        color=(0, 0, 1, 0.6), radius=3, parent=view.scene)

axis = scene.visuals.XYZAxis(parent=view.scene)

# draw and save image
img = canvas.render()
io.write_png("tovispy.png", img)

**Comment by kmuehlbauer** [comment]
@skostrichegg Unfortunately this transparency issues are not sorted out yet. I thought they were. You might playing with different z-levels and using `set_gl_state` with different paramters. 

A quick search revealed these issue # 902, #984. @liubenyuan, are you able to get @skostrichegg on the right path? An other question @rougier, what steps are necessary to finalize this? 

There are several other issues concerning transparency. @Eric89GXL, @davidh-ssec should we create some "transparency super issue" to get hold of all those different issues and to get an impression of the diversity? At least, we could cut down the issue count a bit. Next question is, can we get this sorted out for version 0.5?

**Comment by rougier** [comment]
Transparency is a hard problem because depending on the order you draw things, you can get the right or wrong result. Theoretically, you would have to depth-sort all triangles (which is not do-able for realtime). But there are other techniques that give nice output for most cases. I re-implemented this [paper](http://jcgt.org/published/0002/02/09/paper-lowres.pdf) in glumpy (see https://github.com/glumpy/glumpy/tree/master/examples/transparency).

**Comment by kmuehlbauer** [comment]
@rougier Yes, this is in #984. @liubenyuan implemented a working vispy demo in that issue. My question was aimed on what is necessary to get your glumpy-implementation (using @liubenyuan's demo) in vispy and to make it usable also in the higher level stuff (eg. scene`)? My knowledge of the vispy insides are not that good enough yet, to estimate the amount of code-work needed.

**Comment by rougier** [comment]
It's a 2 (or 3 depending on the undelying techniques) pass rendering so you need some machinery to perform this two pass rendering. This might be quite complex since you have to make sure to display things properly (with correct blending mode) and this might break other things (antialias for example)

**Comment by kmuehlbauer** [comment]
Thanks @rougier, so this is not a quick addition. 

But anyway, since the transparency issues pop up quite frequently, there should at least be something in the documentation (which is missing so far). Glad to hear other opinions.

**Comment by davidh-ssec** [comment]
@kmuehlbauer I'm all for combining all transparency issues in to one. Maybe we can try to clean up some of this or make at least make it easier for users to get it working in a 0.6 release.

**Comment by skostrichegg** [comment]
Thank you all for comment.

I have to adjusting transparency of front one to about 50% (to looks like overlaped).
So, just moving position of object doesn't work for me.

For organizing, I should ask you few more questions.
1. In my understanding, this transparency issue will be fixed in a 0.6.0, is it right?
(And can you give me a general idea of when is the target date for releasing a version for fixed transparency issue?.. )
2. And currently, the way to solve this problem is working with glumpy not vispy?

And this is additional question..
3. @kmuehlbauer mentioned 'high level'.. what does it mean exactly?

Thanks.

**Comment by kmuehlbauer** [comment]
@skostrichegg 

Glad you found a workaround. But still correct use of transparency is very much needed.

To your questions:

1. We ( I mean the current maintainers) try with the help of the guidance of the original developers and other contributors to cut a vispy 0.5 version later this year. It is also planned to get vispy releasing in an more frequent cycle. But which timeframes are to be expected has not finally settled on.

2. @rougier has one implemented one approach into glumpy. @liubenyuan converted this into a vispy demo (https://gist.github.com/liubenyuan/3739c343b3e6ca8117f7). If you try this approach your comments on that are very much appreciated here.
